### PR TITLE
Adding --deployment-container-image-name option to function apps

### PR DESCRIPTION
**Pull request by ahmelsayed**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

### Command Guidelines

- [ ] Each command and parameter has a meaningful description.
- [x] Each new command has a test.

(see [Authoring Command Modules](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules))

### CODE REVIEW

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/tests/test_webapp_commands.py [hunk 1]
```diff
@@ -764,6 +764,30 @@ def test_functionapp_consumption_e2e(self, resource_group, storage_account):
 
         self.cmd('functionapp delete -g {} -n {}'.format(resource_group, functionapp_name))
 
+class FunctionAppOnLinux(ResourceGroupVCRTestBase):
```
Comment (troydai) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/tests/test_webapp_commands.py]: Please do not use `ResourceGroupVCRTestBase` any more.

**Comment by ahmelsayed** [comment]
@yugangw-msft, @troydai 

I added a test case. I can't run it though as I can't run any of your tests following the instructions on https://github.com/Azure/azure-cli/blob/master/doc/configuring_your_machine.md 

from the command line I get 

```powershell
(env) PS ahmels@AHMELS-DEV3 D:\repos\azure-cli
>  echo $env:PYTHONPATH
C:\repos\azure-cli\src

(env) PS ahmels@AHMELS-DEV3 D:\repos\azure-cli
>  .\tools\scripts\run_tests

Directory 'D:\\repos\\azure-cli\\tools\\scripts\\' is not installable. File 'setup.py' not found.
D:\repos\azure-cli\env\Scripts\python.exe: Error while finding module specification for 'automation.tests.run' (ModuleNotFoundError: No module named 'automation.tests')
```

from VS I get

```
Message: Failed to import test module: test_webapp_commands
Traceback (most recent call last):
  File "C:\Users\ahmels\AppData\Local\Programs\Python\Python36-32\lib\unittest\loader.py", line 153, in loadTestsFromName
    module = __import__(module_name)
  File "d:\repos\azure-cli\src\command_modules\azure-cli-appservice\azure\cli\command_modules\appservice\tests\test_webapp_commands.py", line 11, in <module>
    from azure.cli.testsdk import ScenarioTest, LiveScenarioTest, ResourceGroupPreparer, StorageAccountPreparer
  File "d:\repos\azure-cli\src\azure-cli-testsdk\azure\cli\testsdk\__init__.py", line 6, in <module>
    from azure_devtools.scenario_tests import live_only, record_only, get_sha1_hash
ModuleNotFoundError: No module named 'azure_devtools'
```

for all the tests. I don't know how to tell VS to use my venv. I'm opening it from the shell where I have my venv activated, not sure what more I need to do

**Comment by troydai** [comment]
Please run `run_tests` command directly, without referencing the script.

**Comment by troydai** [comment]
@yugangw-msft can you help @ahmelsayed with automation?

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/tests/test_webapp_commands.py (hunk 1 above)
Comment (ahmelsayed) [reply src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/tests/test_webapp_commands.py]: I just copied another test. Is there a doc somewhere that talks about writing tests or a sample test I can look at?

**Comment by ahmelsayed** [comment]
okay, it seems to work now after re-running `dev_setup.py` for some reason. I'll fix the lint issues

**Comment by ahmelsayed** [comment]
* Updated test case to use `ScenarioTest`
* commit a recording
* fix lint issues

**Comment by ahmelsayed** [comment]
oh, I didn't know you also target 2.7 as well. I pushed a fix for that.

**Comment by ahmelsayed** [comment]
also it looks like the CI uses 3.6, while the dev guide calls for 3.5

**Comment by troydai** [comment]
3.5 and 3.6 are very much alike bar from a few new features in 3.6. I will take another look at the pr.

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/tests/test_webapp_commands.py [hunk 2]
```diff
@@ -765,6 +765,27 @@ def test_functionapp_consumption_e2e(self, resource_group, storage_account):
         self.cmd('functionapp delete -g {} -n {}'.format(resource_group, functionapp_name))
 
 
+class FunctionAppOnLinux(ScenarioTest):
+    @ResourceGroupPreparer(location='japanwest')
```
Comment (troydai) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/tests/test_webapp_commands.py]: Why `japanwest`? The recommended location is `southcentralus`, which has lower traffic but is much closer to us.

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py [hunk 3]
```diff
@@ -1433,28 +1435,45 @@ def create_function(resource_group_name, name, storage_account, plan=None,
...
         if is_valid_resource_id(plan):
             plan = parse_resource_id(plan)['name']
         plan_info = client.app_service_plans.get(resource_group_name, plan)
         location = plan_info.location
+        is_linux = plan_info.reserved
+        if is_linux:
+            functionapp_def.kind = 'functionapp,linux'
+            site_config.app_settings.append(NameValuePair('FUNCTIONS_EXTENSION_VERSION', 'beta'))
+            site_config.app_settings.append(NameValuePair('MACHINEKEY_DecryptionKey',
+                                                          str(binascii.hexlify(os.urandom(32)).decode()).upper()))
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: if the goal is just to generate a unique hex string, and there is no hard length requirement involved, maybe you can just go with `uuid.uuid4().hex`, which can by-pass the python 2 and 3 string/bytes difference the code is trying to handle here.

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/tests/test_webapp_commands.py (hunk 2 above)
Comment (ahmelsayed) [reply src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/tests/test_webapp_commands.py]: thanks, done!

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py (hunk 3 above)
Comment (ahmelsayed) [reply src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: it has to be at least 32 byte encryption key in hex format. I don't know if uuid can be used for crypto values, I just read the docs for [`os.urandom`](https://docs.python.org/3/library/os.html#os.urandom) and it seemed to do what I want. Anyway, this seems to work as I'd expect on 2.7 and 3.x, but thanks for the pointer, i'll look at uuid.

**Comment by ahmelsayed** [comment]
@troydai @yugangw-msft thank you for the PR feedback. Hopefully you can merge this today? we need it for connect() and my understanding is that today is the cut off date for cli changes.

**Comment by troydai** [comment]
The cutoff date is 11/8. But we appreciate early bird :)

Please make sure CI passes.

File: src/azure-cli/HISTORY.rst [hunk 4]
```diff
@@ -9,6 +9,7 @@ Release History
 
 2.0.20
 ++++++
+* Allow creating Linux function apps
```
Comment (yugangw-msft) [inline src/azure-cli/HISTORY.rst]: can you back out this change? Rather you should update [app service's history](https://github.com/Azure/azure-cli/blob/dev/src/command_modules/azure-cli-appservice/HISTORY.rst)

Comment (ahmelsayed) [reply src/azure-cli/HISTORY.rst]: oh, sorry I just searched for the first history.rst, didn't know there were module specific ones. Fixed.

**Comment by ahmelsayed** [comment]
@troydai @yugangw-msft I believe I addressed all the comments. Please let me know if there is something I need to do for the merge.

**Pull request merged**